File: Light/led.py
import time
import random
import board
import adafruit_ws2801
import sys
#import colorsys
#import math 

from pynput import keyboard
from enum import Enum

odata = board.MOSI
oclock = board.SCLK
numleds = 96
bright = 1.0
leds = adafruit_ws2801.WS2801(
    oclock, odata, numleds, brightness=bright, auto_write=False
)


# Enum fuer verschiedene Emotionen 
class Emotion(Enum):
    NEUTRAL = 1
    HAPPY = 2
    SAD = 3
    ANGRY = 4
    DISGUSTED = 5
    FEARFUL = 6
    SURPRISED = 7

# Angabe, ob die LEDs ausgeschaltet werden sollen 
notFinish = True 
# aktuelle Emotion
emotion = Emotion.NEUTRAL
# Anzahl der LEDs
n_leds = len(leds)


# Sorgt dafuer, dass die entspr. LEDs beleuchtet werden, abhaengig von
# der uebergebenen Emotion
def illuminateLights():
    # jeder Emotion wird eine Funktion fuer die Beleuchtung zugeordnet
    switcher = {
        Emotion.NEUTRAL: illuminateNeutral,
        Emotion.HAPPY: illuminateHappy,
        Emotion.SAD: illuminateSad,
        Emotion.ANGRY: illuminateAngry,
        Emotion.DISGUSTED: illuminateDisgusted,
        Emotion.FEARFUL: illuminateFearful,
        Emotion.SURPRISED: illuminateSurprised
    }
    # entspr. Funktion je nach Emotion aufrufen
    switcher[emotion]()

# Funktion, um das Licht fuer die neutrale Emotion anzupassen
def illuminateNeutral():
    #leds.fill((138, 30, 0))
    tailCircleAnimation(wait=0.01)


# Funktion, um das Licht fuer die glueckliche Emotion anzupassen
def illuminateHappy():
    #leds.fill((224, 80, 0))
    rainbowCycle()
    leds.show()
    
# Funktion, um das Licht fuer die traurige Emotion anzupassen
def illuminateSad():
    leds.fill((80, 0, 224))
    leds.show()

# Funktion, um das Licht fuer die wuetende Emotion anzupassen
def illuminateAngry():
    leds.fill((224, 0, 0))
    leds.show()

# Funktion, um das Licht fuer die angewiderte Emotion anzupassen
def illuminateDisgusted():
    leds.fill((180, 0, 50))
    leds.show()

# Funktion, um das Licht fuer die aengstliche Emotion anzupassen
def illuminateFearful():
    leds.fill((0, 150, 20))
    leds.show()

# Funktion, um das Licht fuer die ueberraschte Emotion anzupassen
def illuminateSurprised():
    leds.fill((50, 50, 120))
    leds.show()

# Callback fuer einen Tastendruck
def on_release(key):
    # Beleuchtung der LEDs wird beendet
    if key == keyboard.Key.esc:
        global notFinish 
        notFinish = False
        # Stop listener
        return False
    # beim Druck der Leertaste wird zur naechsten Emotion gewechselt
    if key == keyboard.Key.space:
        global emotionChanged
        emotionChanged = True
        
        global emotion
        if emotion.value == 7:
            emotion = Emotion.NEUTRAL
        else:
            emotion = Emotion(emotion.value + 1)
        
        print(emotion)
        return True



listener = keyboard.Listener(
    on_release=on_release)
listener.start()

#######################################################

# Funktion, um alle LEDs auszuschalten
def clear():
    for i in range(n_leds):
        leds[i] = (0, 0, 0)

# Ueberprueft, ob alle LEDs ausgeschaltet sind
def allBlack():
    allBlack = True
    i = 0
    while allBlack and i < n_leds:
        if leds[i] != (0,0,0):
            allBlack = False
        i += 1
    return allBlack

 
#Funktion zur Animation eines "Schweifes"
def tailCircleAnimation(wait=0.01):
    lightingLeds = 9
    leds[0] = (138, 30, 0)
    i = 1
    while (not allBlack()) and notFinish:
        if i < n_leds:
            leds[i] = (138, 30, 0)
            
        if i >= lightingLeds:
            idx = min(n_leds - 1, i - lightingLeds)
            step = 5
            while idx >= 0:
                value = leds[idx]
                r = int(max(0, value[0] - step))
                g = int(max(0, value[1] - step))
                b = int(max(0, value[2] - step))
                leds[idx] = ( r, g, b )
                idx -= 1
        
        leds.show()
        if wait > 0:
            time.sleep(wait)
        i += 1

 # Define the wheel function to interpolate between different hues.
def wheel(pos):
    value = (0,0,0)
    if pos < 85:
        value = (pos * 3, 255 - pos * 3, 0)
    elif pos < 170:
        pos -= 85
        value = (255 - pos * 3, 0, pos * 3)
    else:
        pos -= 170
        value = (0, pos * 3, 255 - pos * 3)
    return value

def rainbowCycle(wait=0.005):
    for j in range(256): # one cycle of all 256 colors in the wheel
        for i in range(n_leds):
            leds[i] = wheel(((i * 256 // n_leds) + j) % 256) 
        leds.show()
        if wait > 0:
            time.sleep(wait)

######################### MAIN LOOP ##############################

while notFinish:
    # LEDs beleuchten 
    illuminateLights()
    clear()
    

leds.deinit()
sys.exit()

File: Light/led.c
#include <FastLED.h>

// TODO Pins anpassen
#define LED_PIN     5
#define NUM_LEDS    96
#define BRIGHTNESS  255
#define LED_TYPE    WS2801
#define COLOR_ORDER RGB
CRGB leds[NUM_LEDS];

#define UPDATES_PER_SECOND 100

/**
 * Angabe, ob Animatation angezeigt werden soll
 */
static int animationOn = 1;

/**
 * Beeinhaltet die einzelnen Emotionen
 */
typedef enum {
    NEUTRAL, HAPPY, SAD, ANGRY, DISGUSTED, FEARFUL, SURPRISED
} Emotion;


static void DrawComet()
{
    const byte fadeAmt = 128;
    const int cometSize = 5;
    const int deltaHue  = 4;

    static byte hue = HUE_RED;
    static int iDirection = 1;
    static int iPos = 0;

    hue += deltaHue;

    iPos += iDirection;
    if (iPos == (NUM_LEDS - cometSize) || iPos == 0)
        iDirection *= -1;
    
    for (int i = 0; i < cometSize; i++)
        leds[iPos + i].setHue(hue);
    
    // Randomly fade the LEDs
    for (int j = 0; j < NUM_LEDS; j++)
        if (random(10) > 5)
            leds[j] = leds[j].fadeToBlackBy(fadeAmt);  

    delay(30);
}

void DrawTwinkle3()
{
    
    leds[random(NUM_LEDS)] = TwinkleColors[random(0, NUM_COLORS)];
        // Randomly fade the LEDs
    for (int j = 0; j < NUM_LEDS; j++)
            leds[j] = leds[j].fadeToBlackBy(64);
    delay(100);       
}



static void illuminateNeutral() {
    if (animationOn) {
        DrawComet();
    }
    else {
        fill_solid(leds, NUM_LEDS, CRGB:RED);
    }
}

void setup() {
  Serial.begin(9600);
  FastLED.addLeds<LED_TYPE, LED_PIN, COLOR_ORDER>(leds, NUM_LEDS).setCorrection( TypicalLEDStrip );
  FastLED.setBrightness(  BRIGHTNESS );

}

void loop() {
    if (Serial.available()) {}
        /* TODO korrektes Einlesen */
        Emotion emotion = (Serial.read() - '0');
        Serial.println("Emotion received");
        
        /*
            in Python dann:
            import serial
            ser = serial.Serial('/dev/ttyUSB0', 9600)
            while 1: 
                if(ser.in_waiting >0):
                    line = ser.readline()
                    print(line) 
         */
        animationOn = Serial.read();
        //delay(2000);
    }
    /*---------------------------------*/
    switch(emotion) {
        case NEUTRAL:
            illuminateNeutral();
            break;
        case HAPPY:
            illuminateHappy();
            break;
        case SAD:
            illuminateSad();
            break;
        case ANGRY:
            illuminateAngry();
            break;
        case DISGUSTED:
            illuminateDisgusted();
            break;
        case FEARFUL:
            illuminateFearful();
            break;
        case SURPRISED:
            illuminateSurprised();
            break;
    }
}